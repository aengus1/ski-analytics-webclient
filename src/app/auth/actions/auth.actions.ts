import {Action} from '@ngrx/store';
import {Authenticate, ConfirmUser, SignupUser, User} from '../model/user';

export enum AuthActionTypes {
  Login = '[Auth] Login',
  Logout = '[Auth] Logout',
  LoginSuccess = '[Auth] Login Success',
  LoginFailure = '[Auth] Login Failure',
  LoginRedirect = '[Auth] Login Redirect',
  Signup = '[Auth] Signup',
  SignupSuccess  = '[Auth] Signup Success',
  SignupFailure = '[Auth] Signup Failure',
  Confirm = '[Auth] Confirm',
  ConfirmSuccess = '[Auth] ConfirmSuccess',
  ConfirmFailure = '[Auth] ConfirmFailure',
  ResendConfirmCode = '[Auth] Resend Confirm code',
  ResendConfirmCodeSuccess = '[Auth] Resend Confirm Success',
  ResendConfirmCodeFailure = '[Auth] Resend Confirm Failure',
  ForgotPassword = '[Auth] Forgot Password',
  ForgotPasswordSuccess = '[Auth] Forgot Password Success',
  ForgotPasswordFailure= '[Auth] Forgot Password Failure'
}

export class Login implements Action {
  readonly type = AuthActionTypes.Login;

  constructor(public payload: Authenticate) {}
}

export class LoginSuccess implements Action {
  readonly type = AuthActionTypes.LoginSuccess;

  constructor(public payload: { user: User }) {}
}

export class LoginFailure implements Action {
  readonly type = AuthActionTypes.LoginFailure;

  constructor(public payload: any) {
  }
}

export class LoginRedirect implements Action {
  readonly type = AuthActionTypes.LoginRedirect;
}

export class Logout implements Action {
  readonly type = AuthActionTypes.Logout;
}

export class Signup implements Action {
  readonly type = AuthActionTypes.Signup;

  constructor(public payload: SignupUser) {}
}

export class SignupSuccess implements Action {
  readonly type = AuthActionTypes.SignupSuccess;

  constructor(public payload: any) {}
}

export class SignupFailure implements Action {
  readonly type = AuthActionTypes.SignupFailure;

  constructor(public payload: any) {
    console.log('signup failure: ' + payload.message);
  }
}

export class Confirm implements Action {
  readonly type = AuthActionTypes.Confirm;

  constructor(public payload: ConfirmUser) {}
}

export class ConfirmSuccess implements Action {
  readonly type = AuthActionTypes.ConfirmSuccess;

  constructor(public payload: any) {}
}

export class ConfirmFailure implements Action {
  readonly type = AuthActionTypes.ConfirmFailure;

  constructor(public payload: any) {}
}
export class ResendConfirmCode implements Action {
  readonly type = AuthActionTypes.ResendConfirmCode;

  constructor(public payload: SignupUser) {}
}

export class ResendConfirmCodeFailure implements Action {
  readonly type = AuthActionTypes.ResendConfirmCodeFailure;

  constructor(public payload: any) {}
}
export class ResendConfirmCodeSuccess implements Action {
  readonly type = AuthActionTypes.ResendConfirmCodeSuccess;

  constructor(public payload: any) {}
}

export class ForgotPassword implements Action {
  readonly type = AuthActionTypes.ForgotPassword;

  constructor(public payload: string) {}
}

export class ForgotPasswordSuccess implements Action {
  readonly type = AuthActionTypes.ForgotPasswordSuccess;

  constructor(public payload: { confirmCode: string }) {}
}

export class ForgotPasswordFailure implements Action {
  readonly type = AuthActionTypes.ForgotPasswordFailure;

  constructor(public payload: any) {
  }
}





export enum SignupStatus  { NOT_STARTED , SIGNUP_PENDING , SIGNUP_COMPLETE,
CONFIRM_PENDING, CONFIRM_COMPLETE, RESEND_CONFIRM_FAILED }
export type AuthActions =
  | Login
  | LoginSuccess
  | LoginFailure
  | LoginRedirect
  | Logout
  | Signup
  | SignupSuccess
  | SignupFailure
  | Confirm
  | ConfirmSuccess
  | ConfirmFailure
  | ResendConfirmCode
  | ResendConfirmCodeFailure
  | ResendConfirmCodeSuccess
  | ForgotPassword
  | ForgotPasswordFailure
  | ForgotPasswordSuccess
  ;
