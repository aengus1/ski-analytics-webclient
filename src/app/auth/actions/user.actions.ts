import {Action} from '@ngrx/store';
import {
  MutationAddUserArgs,
  MutationSaveGenderArgs,
  MutationSaveHeightArgs,
  MutationSaveHrZonesArgs,
  MutationSaveWeightArgs
} from '../../../generated/graphql';

export enum UserActionTypes {
  GetUser = '[User] Get User',
  AddUserSettings = '[User] Add User Settings',
  SaveUserHeight = '[User] Save User Height',
  SaveUserWeight = '[User] Save User Weight',
  SaveUserHrZones = '[User] Save User Hr Zones',
  SaveUserGender = '[User] Save User Gender',
}

export class GetUser implements Action {
  readonly type = UserActionTypes.GetUser;

  constructor(public payload: string) {}
}

export class AddUserSettings implements Action {
  readonly type = UserActionTypes.AddUserSettings;

  constructor(public payload: MutationAddUserArgs) {}
}

export class SaveUserHeight implements Action {
  readonly type = UserActionTypes.SaveUserHeight;

  constructor(public payload: MutationSaveHeightArgs) {
  }
}

export class SaveUserWeight implements Action {
  readonly type = UserActionTypes.SaveUserWeight;

  constructor(public payload: MutationSaveWeightArgs) {
  }
}

export class SaveUserHrZones implements Action {
  readonly type = UserActionTypes.SaveUserHrZones;

  constructor(public payload: MutationSaveHrZonesArgs) {
  }
}

export class SaveUserGender implements Action {
  readonly type = UserActionTypes.SaveUserGender;

  constructor(public payload: MutationSaveGenderArgs) {
  }
}

export type UserActions =
  | GetUser
  | AddUserSettings
  | SaveUserGender
  | SaveUserWeight
  | SaveUserHeight
  | SaveUserHrZones
  ;
