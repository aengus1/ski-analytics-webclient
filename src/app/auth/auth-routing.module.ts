import {NgModule} from '@angular/core';
import {RouterModule, Routes} from '@angular/router';
import {LoginPageComponent} from './containers/login-page.component';
import {SignupPageComponent} from './containers/signup-page.component';

const routes: Routes = [
  {path: 'signin', component: LoginPageComponent},
  {path: 'signup', component: SignupPageComponent}
  ];
@NgModule({
  imports: [
    RouterModule.forChild(routes)
  ],
  exports: [ RouterModule ]
})
export class AuthRoutingModule { }
