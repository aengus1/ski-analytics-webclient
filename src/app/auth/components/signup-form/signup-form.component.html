<div *ngIf="!signupSuccess">
  <form class="form-signup" novalidate [formGroup]="signupForm" (ngSubmit)="submitSignup()">
    <img class="mb-4" src="../../assets/images/logo_only.png" alt="" width="64" height="64">
    <h1 class="h3 mb-3 font-weight-normal">Register</h1>
    <div class="row">
      <div class="col-md-6">
        <label for="inputFirstName" class="sr-only">First Name</label>
        <input type="text" id="inputFirstName" formControlName="firstName" class="form-control" placeholder="First name"
               autocomplete="given-name"
               required>
      </div>
      <div class="col-md-6">
        <label for="inputLastName" class="sr-only">Last Name</label>
        <input type="text" id="inputLastName" formControlName="lastName" class="form-control" placeholder="Last name"
               autocomplete="family-name"
               required>
      </div>
    </div>

    <div class="row">
      <div class="col-md-6">
        <label for="inputEmail" class="sr-only">Email address</label>
        <input type="email" id="inputEmail" formControlName="username" class="form-control" placeholder="Email address"
               autocomplete="email"
               required>
      </div>
        <div class="col-md-6">
          <label for="inputPassword" class="sr-only">Password</label>
          <input type="password" id="inputPassword" formControlName="password" class="form-control"
                 placeholder="Password"
                 autocomplete="new-password"
                 required>
        </div>
      </div>
      <div class="row">
        <div class="col-md-6">
          <label for="inputPasswordConfirm" class="sr-only">Password Confirm</label>
          <input type="password" id="inputPasswordConfirm" name="passwordConfirm" formControlName="passwordConfirm"
                 class="form-control" placeholder="Password confirm"
                 autocomplete="new-password"
                 required>
        </div>
      </div>

    <button class="btn btn-lg btn-primary btn-block" type="submit" [disabled]="signupForm.invalid">Sign up</button>
    <br>
    <!-- SIGNUP ERROR MESSAGE DISPLAY -->
    <div *ngIf="signupErrorMessage" class="errorMessage">
      {{ signupErrorMessage }}
      <div *ngIf="signupErrorMessage==='An account with the given email already exists.'">
        <button class="btn btn-sm btn-secondary btn-block" type="button" (click)="resendConfirm()">Resend Code</button>
      </div>
    </div>
    <!-- RESEND CONFIRM CODE ERROR MESSAGE DISPLAY -->
    <div *ngIf="resendConfirmErrorMessage" class="errorMessage">
      {{ resendConfirmErrorMessage }}
      <div *ngIf="resendConfirmErrorMessage==='User is already confirmed.'">
        <a routerLink="/forgot">
          <button class="btn btn-sm btn-secondary">Reset Password</button>
        </a>
        <a routerLink="/signin">
          <button class="btn btn-sm btn-secondary">Sign In</button>
        </a>
      </div>
    </div>

    <!-- VALIDATION -->
    <div
      class="alert alert-danger"
      *ngIf="signupForm.get('username').hasError('required') && signupForm.get('username').touched">
      Username is required
    </div>
    <div
      class="alert alert-danger"
      *ngIf="signupForm.get('username').hasError('email') && signupForm.get('username').touched">
      not a valid email address
    </div>
    <!-- password validation error -->
    <div
      class="alert alert-danger"
      *ngIf="signupForm.get('password').hasError('required') && signupForm.get('password').touched">
      Password is required
    </div>
    <div
      class="alert alert-danger"
      *ngIf="signupForm.get('password').hasError('pattern') && signupForm.get('password').touched">
      Password must be at least 8 characters, and contain a mixture of upper case, lower case and number characters
    </div>
    <div
      class="alert alert-danger"
      *ngIf="signupForm.errors?.mismatch && (signupForm.controls['passwordConfirm'].dirty || signupForm.controls['passwordConfirm'].touched)">
      Passwords must match
    </div>
    <div
      class="alert alert-danger"
      *ngIf="signupForm.get('firstName').hasError('minLength') && signupForm.get('firstName').touched">
      First name must be at least 2 characters
    </div>
    <div
      class="alert alert-danger"
      *ngIf="signupForm.get('lastName').hasError('minLength') && signupForm.get('lastName').touched">
      Last name must be at least 2 characters
    </div>

    <hr>
    Already have an account ? <a routerLink="/signin"> Sign In</a>
    <p class="mt-5 mb-3 text-muted">&copy; 2017-2018</p>
  </form>
</div>

<!-- CONFIRM CODE FORM -->
<div *ngIf="signupSuccess">
  <form class="form-confirm" novalidate [formGroup]="confirmForm" (ngSubmit)="submitConfirm()">
    <img class="mb-4" src="../../assets/images/logo_only.png" alt="" width="64" height="64">
    <h1 class="h3 mb-3 font-weight-normal">Enter validation code:</h1>

    <label for="validationCode" class="sr-only">Enter code:</label>
    <input type="text" id="validationCode" formControlName="validationCode" class="form-control" placeholder="Code"
           required>

    <button class="btn btn-lg btn-primary btn-block" type="submit" [disabled]="confirmForm.invalid">Confirm</button>
    <br>

    <!-- ERROR MESSAGE DISPLAY -->
    <div *ngIf="confirmErrorMessage" class="errorMessage">
      {{ confirmErrorMessage }}
      <div *ngIf="confirmErrorMessage==='User is already confirmed.'">
        <a routerLink="/forgot">
          <button class="btn btn-sm btn-secondary">Reset Password</button>
        </a>
        <a routerLink="/signin">
          <button class="btn btn-sm btn-secondary">Sign In</button>
        </a>
      </div>
    </div>

    <!-- VALIDATION -->
    <div
      class="alert alert-danger"
      *ngIf="confirmForm.get('validationCode').hasError('required') && confirmForm.get('validationCode').touched">
      Validation code is required
      <button class="btn btn-sm btn-secondary btn-block" type="button" (click)="resendConfirm()">Resend Code</button>
    </div>


    <hr>
    Already have an account ? <a routerLink="/signin"> Sign In</a>
    <p class="mt-5 mb-3 text-muted">&copy; 2017-2018</p>
  </form>
</div>

<ng-template #confirmSuccessModal>
  <div class="modal-header">
    <h4 class="modal-title">Success</h4>
  </div>
  <div class="modal-body">
    <p>Registration successful</p>
  </div>
  <div class="modal-footer">
    <button type="button" class="btn btn-md btn-primary btn-block" (click)="gotoLogin()">Go to Login</button>
    <button type="button" class="btn btn-md btn-primary" (click)="closeModal()">Close</button>
  </div>
</ng-template>
