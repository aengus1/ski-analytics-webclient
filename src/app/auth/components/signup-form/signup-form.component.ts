import {Component, EventEmitter, Input, OnInit, Output, ViewChild} from '@angular/core';
import {FormControl, FormGroup, Validators} from '@angular/forms';
import {ConfirmUser, SignupUser} from '../../model/user';
import {SignupStatus} from '../../actions/auth.actions';
import {NgbModal, NgbModalOptions, NgbModalRef} from '@ng-bootstrap/ng-bootstrap';
import {Router} from '@angular/router';

@Component({
  selector: 'app-signup-form',
  templateUrl: './signup-form.component.html',
  styleUrls: ['./signup-form.component.scss']
})
export class SignupFormComponent implements OnInit {

  public status: SignupStatus = SignupStatus.NOT_STARTED;
  public signupSuccess = false;
  private signupEmail = '';
  private modalRef: NgbModalRef;
  private modalCloseResult: string;


  @Input()
  set signupStatus(status: SignupStatus) {
    if (status !== SignupStatus.NOT_STARTED) {
      this.signupForm.disable();
    } else {
      this.signupForm.enable();
    }
    if (status !== SignupStatus.SIGNUP_COMPLETE) {
      this.confirmForm.disable();
    } else {
      this.confirmForm.enable();
      this.signupSuccess = true;
    }
    if (status === SignupStatus.CONFIRM_COMPLETE) {
      const options: NgbModalOptions = {
        size: 'lg'
      };
      setTimeout(() => {
          this.modalRef = this.modal.open(this.confirmSuccessModal, options);
          this.modalRef.result.then((result) => {
            this.modalCloseResult = `Closed with: ${result}`;
          }, (reason) => {
            this.modalCloseResult = 'Dismissed';
          });
        }
        , 100);
    }
  }

  @Input() signupErrorMessage: string | null;

  @Input() confirmErrorMessage: string | null;

  @Input() resendConfirmErrorMessage: string | null;

  @Output() submittedSignup = new EventEmitter<SignupUser>();

  @Output() submittedConfirm = new EventEmitter<ConfirmUser>();

  @Output() submittedResendConfirm = new EventEmitter<SignupUser>();

  @ViewChild('confirmSuccessModal', { static: true }) private confirmSuccessModal;

  signupForm: FormGroup = new FormGroup({
    username: new FormControl('', [Validators.required, Validators.email]),
      password: new FormControl('', [
        Validators.required,
        Validators.pattern('^(?=.*[A-Z])(?=.*[a-z])(?=.*\\d)[A-Za-z\\d]{8,20}')
      ]),
      passwordConfirm: new FormControl('',
        [
          Validators.required,
          Validators.pattern('^(?=.*[A-Z])(?=.*[a-z])(?=.*\\d)[A-Za-z\\d]{8,20}')
        ]),
    firstName: new FormControl('', [Validators.required]),
    lastName: new FormControl('', [Validators.required, Validators.minLength(2)])
  }, this.passwordMatchValidator);

  confirmForm: FormGroup = new FormGroup({
    validationCode: new FormControl('', [
      Validators.required
    ])
  });

  constructor(private modal: NgbModal, private router: Router) {
  }

  ngOnInit() {
    this.signupStatus = SignupStatus.NOT_STARTED;
  }

  submitSignup() {
    if (this.signupForm.valid) {
      this.submittedSignup.emit(this.signupForm.value);
      this.signupEmail = this.signupForm.value.username;
    }
  }

  submitConfirm() {
    if (this.confirmForm.valid) {
      this.submittedConfirm.emit({username: this.signupEmail, confirmCode: this.confirmForm.value.validationCode});
    }
  }

  resendConfirm() {
    this.submittedResendConfirm.emit(this.signupForm.value);
  }

  /* edge case when user wants to sign up again.. */
  closeModal() {
    this.modalRef.close();
  }

  gotoLogin() {
    this.modalRef.close();
    this.router.navigateByUrl('/signin');
  }

  passwordMatchValidator(frm: FormGroup) {
    const pass = frm.get('password').value;
    const confirmPass = frm.get('passwordConfirm').value;

    return pass === confirmPass ? null : { mismatch: true };
  }
}

