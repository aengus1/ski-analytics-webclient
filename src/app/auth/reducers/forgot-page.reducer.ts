import {AuthActions, AuthActionTypes} from './../actions/auth.actions';

export interface State {
  error: string | null;
  pending: boolean;
}

export const initialState: State = {
  error: null,
  pending: false,
};

export function reducer(state = initialState, action: AuthActions): State {
  switch (action.type) {
    case AuthActionTypes.ForgotPassword: {
      return {
        ...state,
        error: null,
        pending: true,
      };
    }

    case AuthActionTypes.ForgotPasswordSuccess: {
      return {
        ...state,
        error: null,
        pending: false,
      };
    }

    case AuthActionTypes.ForgotPasswordFailure: {
      return {
        ...state,
        error: action.payload.message,
        pending: false,
      };
    }

    default: {
      return state;
    }
  }
}

export const getError = (state: State) => state.error;
export const getPending = (state: State) => state.pending;
