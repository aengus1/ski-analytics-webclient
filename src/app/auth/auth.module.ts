import {ModuleWithProviders, NgModule} from '@angular/core';
import {CommonModule} from '@angular/common';
import {FormsModule, ReactiveFormsModule} from '@angular/forms';
import {EffectsModule} from '@ngrx/effects';
import {LoginFormComponent} from './components/login-form/login-form.component';
import {LoginPageComponent} from './containers/login-page.component';
import {ForgotFormComponent} from './components/forgot-form/forgot-form.component';
import {ForgotPageComponent} from './containers/forgot-page.component';
import {AuthService} from './services/auth.service';
import {AuthRoutingModule} from './auth-routing.module';
import {StoreModule} from '@ngrx/store';
import {AuthGuard} from './guards/auth.guard';
import {AuthEffects} from './effects/auth.effects';
import {reducers} from './reducers';
import {SignupPageComponent} from './containers/signup-page.component';
import {SignupFormComponent} from './components/signup-form/signup-form.component';
import {RouterModule} from '@angular/router';
import {SharedModule} from '../shared/shared.module';


export const COMPONENTS = [
  LoginPageComponent,
  LoginFormComponent,
  SignupPageComponent,
  SignupFormComponent,
  ForgotPageComponent,
  ForgotFormComponent
];

@NgModule({
  imports: [
    CommonModule, ReactiveFormsModule, FormsModule, RouterModule, SharedModule
  ],
  declarations:  COMPONENTS,
  exports: COMPONENTS
})
export class AuthModule {
  static forRoot(): ModuleWithProviders {
    return {
      ngModule: RootAuthModule,
      providers: [AuthService, AuthGuard ]
    };
  }
}

@NgModule({
  imports: [
    AuthModule,
    AuthRoutingModule,
    StoreModule.forFeature('auth', reducers),
    EffectsModule.forFeature([AuthEffects]),
  ],
  declarations: [],
  exports: []
})
export class RootAuthModule {}
