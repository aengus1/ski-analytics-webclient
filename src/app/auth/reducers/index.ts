import {ActionReducerMap, createFeatureSelector, createSelector} from '@ngrx/store';
import * as fromAuth from './auth.reducer';
import * as fromLoginPage from './login-page.reducer';
import * as fromSignupPage from './signup-page.reducer';
import * as fromForgotPage from './forgot-page.reducer';
import * as fromRoot from '../../reducers';

export interface AuthState {
  status: fromAuth.State;
  loginPage: fromLoginPage.State;
  signupPage: fromSignupPage.State;
  forgotPage: fromForgotPage.State;
}

export interface State extends fromRoot.State {
  auth: AuthState;
}

export const reducers: ActionReducerMap<AuthState> = {
  status: fromAuth.reducer,
  loginPage: fromLoginPage.reducer,
  signupPage: fromSignupPage.reducer,
  forgotPage: fromForgotPage.reducer
};

export const selectAuthState = createFeatureSelector<AuthState>('auth');

export const selectAuthStatusState = createSelector(
  selectAuthState,
  (state: AuthState) => state.status
);
export const getLoggedIn = createSelector(
  selectAuthStatusState,
  fromAuth.getLoggedIn
);

export const getUser = createSelector(selectAuthStatusState, fromAuth.getUser);

export const selectLoginPageState = createSelector(
  selectAuthState,
  (state: AuthState) => state.loginPage
);
export const getLoginPageError = createSelector(
  selectLoginPageState,
  fromLoginPage.getError
);
export const getLoginPagePending = createSelector(
  selectLoginPageState,
  fromLoginPage.getPending
);

export const selectSignupPageState = createSelector(
  selectAuthState,
  (state: AuthState) => state.signupPage
);
export const getSignupPageSignupError = createSelector(
  selectSignupPageState,
  fromSignupPage.getSignupError
);
export const getSignupPageStatus = createSelector(
  selectSignupPageState,
  fromSignupPage.getSignupStatus
);

export const getSignupPageConfirmError = createSelector(
  selectSignupPageState,
  fromSignupPage.getConfirmError
);

export const getSignupPageResendConfirmError = createSelector(
  selectSignupPageState,
  fromSignupPage.getResendConfirmError
);

export const selectForgotPageState = createSelector(
  selectAuthState,
  (state: AuthState) => state.forgotPage
);
export const getForgotPageError = createSelector(
  selectForgotPageState,
  fromForgotPage.getError
);
export const getForgotPagePending = createSelector(
  selectLoginPageState,
  fromForgotPage.getPending
);

