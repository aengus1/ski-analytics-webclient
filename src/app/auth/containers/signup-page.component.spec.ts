import {ComponentFixture, TestBed} from '@angular/core/testing';
import {ReactiveFormsModule} from '@angular/forms';
import {combineReducers, Store, StoreModule} from '@ngrx/store';
import * as fromAuth from '../reducers';
import {SignupFormComponent} from '../components/signup-form/signup-form.component';
import {Confirm, Signup} from '../actions/auth.actions';
import {SignupPageComponent} from './signup-page.component';
import {NgbModal, NgbModule} from '@ng-bootstrap/ng-bootstrap';
import {RouterTestingModule} from '@angular/router/testing';
import {HttpClient} from '@angular/common/http';

describe('SignupPageComponent', () => {
  let fixture: ComponentFixture<SignupPageComponent>;
  let store: Store<fromAuth.State>;
  let instance: SignupPageComponent;
  let modal: NgbModal;

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [
        StoreModule.forRoot({
          auth: combineReducers(fromAuth.reducers),
        }),
        NgbModule,
        ReactiveFormsModule, RouterTestingModule
      ],
      declarations: [SignupPageComponent, SignupFormComponent],
      providers: [NgbModal, HttpClient]
    });
  });
  beforeEach(() => {
    fixture = TestBed.createComponent(SignupPageComponent);
    instance = fixture.componentInstance;
    store = TestBed.inject(Store);
    modal = TestBed.inject(NgbModal);
    spyOn(store, 'dispatch').and.callThrough();
  });

  /**
     * Container components are used as integration points for connecting
     * the store to presentational components and dispatching
     * actions to the store.
     *
     * Container methods that dispatch events are like a component's output observables.
     * Container properties that select state from store are like a component's input properties.
     * If pure components are functions of their inputs, containers are functions of state
     *
     * Traditionally you would query the components rendered template
     * to validate its state. Since the components are analogous to
     * pure functions, we take snapshots of these components for a given state
     * to validate the rendered output and verify the component's output
     * against changes in state.
     */
    it('should compile', () => {
      fixture.detectChanges();

      // noinspection TypeScriptUnresolvedFunction
      (<any>expect(fixture)).toMatchSnapshot();
    });

    it('should dispatch a signup event on submit', () => {
      const $event: any = {};
      const action = new Signup($event);

      instance.onSubmitSignup($event);

      expect(store.dispatch).toHaveBeenCalledWith(action);
    });

  it('should dispatch a confirm event on submit', () => {
    const $event: any = {};
    const action = new Confirm($event);

    instance.onSubmitConfirm($event);

    expect(store.dispatch).toHaveBeenCalledWith(action);
  });
});

