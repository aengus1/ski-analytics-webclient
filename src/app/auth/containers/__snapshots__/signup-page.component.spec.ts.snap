// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignupPageComponent should compile 1`] = `
<app-signup-page
  errorConfirm$={[Function Store]}
  errorSignup$={[Function Store]}
  resendConfirm$={[Function Store]}
  signupStatus$={[Function Store]}
  store={[Function Store]}
>
  <app-signup-form
    ng-reflect-signup-status="0"
  >
    <div
      class="header"
    >
      <button
        class="btn btn-lg btn-primary logo-button "
      >
        <a
          class="logo-font"
          href="http://crunch.ski"
        >
           Crunch.Ski
        </a>
      </button>
    </div>
    
    <div>
      <form
        class="form-signup ng-untouched ng-pristine ng-invalid"
        ng-reflect-form="[object Object]"
        novalidate=""
      >
        <img
          alt=""
          class="mb-4"
          height="64"
          src="../../assets/images/logo_only.png"
          width="64"
        />
        <h1
          class="h3 mb-3 font-weight-normal"
        >
          Register
        </h1>
        <div
          class="row"
        >
          <div
            class="col-md-6"
          >
            <label
              class="sr-only"
              for="inputFirstName"
            >
              First Name
            </label>
            <input
              autocomplete="given-name"
              class="form-control ng-untouched ng-pristine ng-invalid"
              formcontrolname="firstName"
              id="inputFirstName"
              ng-reflect-name="firstName"
              ng-reflect-required=""
              placeholder="First name"
              required=""
              type="text"
            />
          </div>
          <div
            class="col-md-6"
          >
            <label
              class="sr-only"
              for="inputLastName"
            >
              Last Name
            </label>
            <input
              autocomplete="family-name"
              class="form-control ng-untouched ng-pristine ng-invalid"
              formcontrolname="lastName"
              id="inputLastName"
              ng-reflect-name="lastName"
              ng-reflect-required=""
              placeholder="Last name"
              required=""
              type="text"
            />
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-md-6"
          >
            <label
              class="sr-only"
              for="inputEmail"
            >
              Email address
            </label>
            <input
              autocomplete="email"
              class="form-control ng-untouched ng-pristine ng-invalid"
              formcontrolname="username"
              id="inputEmail"
              ng-reflect-name="username"
              ng-reflect-required=""
              placeholder="Email address"
              required=""
              type="email"
            />
          </div>
          <div
            class="col-md-6"
          >
            <label
              class="sr-only"
              for="inputPassword"
            >
              Password
            </label>
            <input
              autocomplete="new-password"
              class="form-control ng-untouched ng-pristine ng-invalid"
              formcontrolname="password"
              id="inputPassword"
              ng-reflect-name="password"
              ng-reflect-required=""
              placeholder="Password"
              required=""
              type="password"
            />
          </div>
        </div>
        <div
          class="row"
        >
          <div
            class="col-md-6"
          >
            <label
              class="sr-only"
              for="inputPasswordConfirm"
            >
              Password Confirm
            </label>
            <input
              autocomplete="new-password"
              class="form-control ng-untouched ng-pristine ng-invalid"
              formcontrolname="passwordConfirm"
              id="inputPasswordConfirm"
              name="passwordConfirm"
              ng-reflect-name="passwordConfirm"
              ng-reflect-required=""
              placeholder="Password confirm"
              required=""
              type="password"
            />
          </div>
        </div>
        <button
          class="btn btn-lg btn-primary btn-block"
          disabled=""
          type="submit"
        >
          Sign up
        </button>
        <br />
        
        
        
        
        
        
        
        
        
        <hr />
         Already have an account ? 
        <a
          href="/signin"
          ng-reflect-router-link="/signin"
          routerlink="/signin"
        >
           Sign In
        </a>
        <p
          class="mt-5 mb-3 text-muted"
        >
          Â© 2020-2021
        </p>
      </form>
    </div>
    
    
  </app-signup-form>
</app-signup-page>
`;
